Unified ModeMap Type

lch opened the pull request:
This PR renames the type `OccupiedModeMap` to `ModeMap`.

## New features
* `occupied_mode_map` and `unoccupied_mode_map`— new constructors for type `ModeMap`. 
* `FermiUnoccupiedModes`— new iterator version generator of type `ModeMap` for `FermiFS`.

## Deprecated features
* `OccupiedModeMap` is deprecated and replaced by `occupied_mode_map`. The type is replaced by `ModeMap`.

Originally, there was a `OccupiedModeMap` type along with its constructor function, which was used to generate a statically allocated array for storing `*FSIndex` values corresponding to various Fock space addresses. However, in some cases, it is also necessary to represent and store unoccupied modes—particularly when working with Fermi addresses, where the occupation number (occnum) is limited to 0 or 1.
The constructors enable consistent handling of both occupied and unoccupied modes. Then the original `OccupiedModeMap` constructor function is deprecated with `Base.@deprecate`.

lch committed:
Unified ModeMap Type (#330)

1. Renamed the type OccupiedModeMap to ModeMap and replaced all occurrences.
2. Added the occupied_mode_map function as a constructor for ModeMap.
3. Added the unoccupied_mode_map function as a constructor for FermiFS
and corresponding iterator type FermiUnoccupiedModes.

joachimbrand commented on src/BitStringAddresses/fockaddress.jl:
```diff
@@ -209,35 +209,40 @@ See [`SingleComponentFockAddress`](@ref).
+
 """
-    OccupiedModeMap(addr) <: AbstractVector
+    ModeMap(addr) <: AbstractVector
```
```suggestion
    ModeMap <: AbstractVector
```
We are only documenting a type, not a constructor. Hence we should leave out the `()`, which would indicate that you can use it as a function/constructor.

joachimbrand commented on src/BitStringAddresses/fockaddress.jl:
```diff
@@ -209,35 +209,40 @@ See [`SingleComponentFockAddress`](@ref).
+It is similar to [`onr`](@ref) but contains more information.
+
+Do not construct this directly. Use [`occupied_mode_map`](@ref) or 
+[`unoccupied_mode_map`](@ref) to obtain an instance.
```
Same comment here. Just say that a `ModeMap` is returned by `occupied_mode_map` and `unoccupied_mode_map`.

joachimbrand reviewed (commented):
I guess now you'll have to do a "search and replace" in many files in the `Hamiltonians` module.

joachimbrand commented on src/BitStringAddresses/fermifs.jl:
```diff
@@ -131,6 +131,50 @@ Base.:(==)(a::FermiFS, b::FermiFS) = a.bs == b.bs
+`unoccupied_mode_map(addr)[i]` contains the index for the `i`-th unoccupied mode.
+This is useful because unoccupied modes is required in some cases.
+`unoccupied_mode_map(addr)` is an eager version of the iterator returned by
+[`unoccupied_modes`](@ref). It is similar to [`onr`](@ref) but contains more information.
```
```suggestion
[`unoccupied_modes`](@ref). It is similar to 
[`occupation_number_representation`](@ref) but contains more information.
```

joachimbrand commented on src/BitStringAddresses/fermifs.jl:
````diff
@@ -131,6 +131,50 @@ Base.:(==)(a::FermiFS, b::FermiFS) = a.bs == b.bs
+true
+
+```
+See also [`SingleComponentFockAddress`](@ref).
````
```suggestion
See also [`occupied_mode_map`](@ref).
```

joachimbrand commented on src/BitStringAddresses/fockaddress.jl:
````diff
@@ -176,6 +176,27 @@ See also [`find_occupied_mode`](@ref),
+FermiFSIndex(occnum=0, mode=5, offset=4)
+FermiFSIndex(occnum=0, mode=6, offset=5)
+```
+See also [`find_occupied_mode`](@ref), [`SingleComponentFockAddress`](@ref).
````
```suggestion
See also [`find_occupied_mode`](@ref), [`occupied_modes`](@ref).
```

joachimbrand commented on src/BitStringAddresses/fockaddress.jl:
````diff
@@ -255,12 +299,7 @@ julia> dot(mf, 1:20)
 ```
 See also [`dot`](@ref Main.Hamiltonians.dot), [`SingleComponentFockAddress`](@ref).
````
```suggestion
See also [`dot`](@ref Main.Hamiltonians.dot), [`unoccupied_mode_map`](@ref),
[`SingleComponentFockAddress`](@ref).
```

joachimbrand commented on src/BitStringAddresses/fockaddress.jl:
```diff
@@ -209,16 +230,39 @@ See [`SingleComponentFockAddress`](@ref).
+This struct is not intended to construct directly. Use [`occupied_mode_map`](@ref) or 
+[`unoccupied_mode_map`](@ref) to obtain an instance.
+
+See also [`dot`](@ref Main.Hamiltonians.dot), [`SingleComponentFockAddress`](@ref).
```
```suggestion
See also [`SingleComponentFockAddress`](@ref).
```

joachimbrand commented on src/BitStringAddresses/BitStringAddresses.jl:
```diff
@@ -18,8 +18,8 @@ export OccupationNumberFS
-export find_occupied_mode, find_mode, occupied_modes, is_occupied, num_occupied_modes
-export excitation, near_uniform, OccupiedModeMap, OccupiedPairsMap
+export find_occupied_mode, find_mode, occupied_modes, unoccupied_modes, is_occupied, num_occupied_modes, num_unoccupied_modes
+export excitation, near_uniform, ModeMap, OccupiedPairsMap, occupied_mode_map, unoccupied_mode_map
```
```suggestion
export excitation, near_uniform, OccupiedPairsMap, occupied_mode_map, unoccupied_mode_map
```
I lean towards not exporting `ModeMap`, as the user typically would not have to interface with it.

joachimbrand replied:
This will require explicitly importing the symbol in `Hamiltonians/Hamiltonians.jl`:
```julia
using ..BitStringAddresses: ModeMap
```

joachimbrand reviewed (changes requested):
Thanks for the PR!
It looks very good. Please see comments below!

lch committed:
Unified ModeMap Type (#330)

1. Apply suggestions from code review.
2. Import ModeMap in Hamiltonians module explicitly.

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

mtsch commented on src/BitStringAddresses/fockaddress.jl:
```diff
@@ -209,16 +230,39 @@ See [`SingleComponentFockAddress`](@ref).
+It stores the FSIndex of corresponding address as an `AbstractVector` compatible with
+[`excitation`](@ref) - [`BoseFSIndex`](@ref) or [`FermiFSIndex`](@ref).
+
+This struct is not intended to construct directly. Use [`occupied_mode_map`](@ref) or 
```
```suggestion
This struct is not intended to be constructed directly. Use [`occupied_mode_map`](@ref) or 
```

mtsch commented on test/BitStringAddresses.jl:
```diff
@@ -549,7 +549,7 @@ end
         @test onr(lfs, LadderBoundaries(2, 3)) == [1 0 0; 1 1 0]
         @test num_occupied_modes(lfs) == length(occupied_modes(lfs)) == 3
-        @test OccupiedModeMap(lfs) == collect(occupied_modes(lfs))
+        @test occupied_mode_map(lfs) == collect(occupied_modes(lfs))
```
Hmm I realized we don't have a test like this for `FermiFS` or `BoseFS`. Could you maybe sneak it in somewhere?

mtsch reviewed (changes requested):
Looks very good! A minor comment in one of the docstrings and I would also like to see a test similar to the following test but for unoccupied modes, something like
```julia
    @testset "unoccupied_modes" begin
        for o in (small, big, giant)
            f = FermiFS(o)
            sites = collect(unoccupied_modes(f))
            @test getproperty.(sites, :mode) == findall(==(0), onr(f))
        end
    end
```
in `@testset "FermiFS"`

lch committed:
Unified ModeMap Type (#330)

1. Add tests to new methods according to suggestion

Co-authored-by: mtsch <matijacufar@gmail.com>

lch committed:
Unified ModeMap Type (#330)

1. Fix bug in unoccupied_modes method

lch committed:
Unified ModeMap Type (#330)

1. Resolve unoccupied_modes method test failed when FermiFS is
   constructed with SortedParticleList

joachimbrand commented on src/BitStringAddresses/sortedparticlelist.jl:
```diff
@@ -277,3 +277,23 @@ function Base.iterate(fom::FermiOccupiedModes{<:Any,<:SortedParticleList}, i=1)
     end
 end
+
+import Base.:~
+function ~(ss::SortedParticleList{N,M,T}) where {N,M,T}
+    ss
+end
+Base.length(::FermiUnoccupiedModes{N,<:SortedParticleList}) where {N} = N
```
Please remove the `~` method for `SortedParticleList`. It doesn't make sense. The method for `length` in line 285 is strictly unnecessary, as it is already covered by a method defined in `fockaddress.jl`.
```suggestion
```

joachimbrand commented on src/BitStringAddresses/fockaddress.jl:
```diff
@@ -644,13 +684,24 @@ Base.show(io::IO, ::MIME"text/plain", i::FermiFSIndex) = show(io, i)
 Base.eltype(::FermiOccupiedModes) = FermiFSIndex
 
+"""
+    FermiUnoccupiedModes{N,S<:BitString}
```
```suggestion
    FermiUnoccupiedModes{N}
```
No need to specify the storage type here. The docstring is for the more general type.

lch replied:
Perhaps `FermiOccupiedModes{N,S<:BitString}` could also be changed to maintain consistency.

joachimbrand replied:
I don't understand your comment. Do you mean you also want to change (and generalise) the type definition?

My comment was a suggestion to remove the obsolete reference to `Bitstring` in the doc string for the type. If you agree, you still have to commit the change, as my comment is only a suggestion.

lch replied:
I mean perhaps the docstring for `FermiOccupiedModes` should be updated from
```julia
"""
    FermiOccupiedModes{N,S<:BitString}

Iterator over occupied modes in address. `N` is the number of fermions. See [`occupied_modes`](@ref).
"""
```
to
```julia
"""
    FermiOccupiedModes{N}

Iterator over occupied modes in address. `N` is the number of fermions. See [`occupied_modes`](@ref).
"""
```
as well.

joachimbrand commented on src/BitStringAddresses/fermifs.jl:
```diff
@@ -131,6 +131,54 @@ Base.:(==)(a::FermiFS, b::FermiFS) = a.bs == b.bs
 occupied_modes(a::FermiFS{N,<:Any,S}) where {N,S} = FermiOccupiedModes{N,S}(a.bs)
 
+num_unoccupied_modes(::FermiFS{N,M}) where {N,M} = M - N
+function unoccupied_modes(a::FermiFS{N,M,S}) where {N,M,S}
+    s = ~a.bs
+    FermiUnoccupiedModes{M - N,typeof(s)}(s)
+end
```
This should specialise for the storage type `S` and only invert the bitstring with `~` if the storage type is a bitstring.

mtsch commented on src/BitStringAddresses/sortedparticlelist.jl:
```diff
@@ -277,3 +277,23 @@ function Base.iterate(fom::FermiOccupiedModes{<:Any,<:SortedParticleList}, i=1)
+        return nothing
+    else
+        i, j = state
+        while i ∈ fum.storage.storage
```
Keep in mind that `∈` performs a linear search every time. I checked and it doesn't matter here because the list in a `SortedParticleList` is going to be very short.

I'm just pointing it out because you should be wary of using this pattern elsewhere.

lch replied:
Thanks for the suggestion. I have updated my iterator implementation.

mtsch commented on src/BitStringAddresses/fermifs.jl:
```diff
@@ -131,6 +131,54 @@ Base.:(==)(a::FermiFS, b::FermiFS) = a.bs == b.bs
 
+num_unoccupied_modes(::FermiFS{N,M}) where {N,M} = M - N
+function unoccupied_modes(a::FermiFS{N,M,S}) where {N,M,S}
+    s = ~a.bs
```
I'm not sure I understand why the bitstring needs to be inverted here. Can't you just search for zeros in `iterate` instead of searching for ones in the inverted bitstring? I'm pointing this out because it's not obvious later in the code that the bistring stored as `storage` in `FermiUnoccupiedModes` is inverted.

If searching for ones is easier than searching for zeros, you can also invert the bitstring in `iterate` itself. Inverting an integer is almost free on modern processors and shouldn't affect performance.

lch replied:
Allow me to give it a try.

mtsch commented on src/BitStringAddresses/fermifs.jl:
```diff
@@ -131,6 +131,54 @@ Base.:(==)(a::FermiFS, b::FermiFS) = a.bs == b.bs
+`unoccupied_mode_map(addr)` is an eager version of the iterator returned by
+[`unoccupied_modes`](@ref). It is similar to 
+[`onr`](@ref) but contains more information.
+
```
```suggestion

Note that this function is only implemented for addresses of type [`FermiFS`](@ref).

```
We don't really need this function for bosonic addresses, but I'd add this comment to the docstring just to make it even more explicit.

lch committed:
Unified ModeMap Type (#330)

1. Rewrite the FermiUnoccupiedModes iterator based on the suggestions.
2. Improve the docstring.

joachimbrand reviewed (approved):
Looks great. Thanks for the PR!

mtsch reviewed (approved):
Looks great!
